src/workerd/io/io-context.h

```diff
@@ -1257,23 +1257,40 @@ kj::_::ReducePromises<RemoveIoOwn<T>> IoContext::awaitJs(jsg::Lock& js, jsg::Pro
   auto paf = kj::newPromiseAndFulfiller<RemoveIoOwn<T>>();
   struct RefcountedFulfiller: public Finalizeable, public kj::Refcounted {
     kj::Own<kj::PromiseFulfiller<RemoveIoOwn<T>>> fulfiller;
+    kj::Own<const workerd::Worker::Isolate> isolate;
     bool isDone = false;
 
-    RefcountedFulfiller(kj::Own<kj::PromiseFulfiller<RemoveIoOwn<T>>> fulfiller)
-        : fulfiller(kj::mv(fulfiller)) {}
+    RefcountedFulfiller(kj::Own<const workerd::Worker::Isolate> isolate,
+        kj::Own<kj::PromiseFulfiller<RemoveIoOwn<T>>> fulfiller)
+        : fulfiller(kj::mv(fulfiller)),
+          isolate(kj::mv(isolate)) {}
 
     ~RefcountedFulfiller() noexcept(false) {
       if (!isDone) {
-        // The JavaScript resolver was garbage collected, i.e. JavaScript will never resolve
-        // this promise.
-        fulfiller->reject(JSG_KJ_EXCEPTION(FAILED, Error, "Promise will never complete."));
+        auto hasExcessivelyExceededHeapLimit =
+            isolate->getLimitEnforcer().hasExcessivelyExceededHeapLimit();
+        if (hasExcessivelyExceededHeapLimit) {
+          fulfiller->reject(
+              JSG_KJ_EXCEPTION(OVERLOADED, Error, "jsg.Error: Worker has exceeded memory limit."));
+        } else {
+          // The JavaScript resolver was garbage collected, i.e. JavaScript will never resolve
+          // this promise.
+          fulfiller->reject(JSG_KJ_EXCEPTION(FAILED, Error, "Promise will never complete."));
+        }
       }
     }
 
    private:
     kj::Maybe<kj::StringPtr> finalize() override {
       if (!isDone) {
-        fulfiller->reject(JSG_KJ_EXCEPTION(FAILED, Error, "Promise will never complete."));
+        auto hasExcessivelyExceededHeapLimit =
+            isolate->getLimitEnforcer().hasExcessivelyExceededHeapLimit();
```

Comment:
Nit: the duplicated logic here has always bothered me. Any chance we could combine these into a single utility method and eliminate/reduce the duplication?